### ISSUE TITLE
`transform_inclusive_scan` gives incorrect results for non-commutative operator

### ISSUE THREAD

**Issue by LiliumAtratum** [STEllAR-GROUP/hpx#4787]
## Expected Behavior

`transform_inclusive_scan` should work correctly when the `op` is a non-commutative operator. 

Consider the following example:
Trying the perform an inclusive prefix scan of array: [1, 10, 100, 1000]
with a binary op(x,y) defined as follows:
if x<y then x+y else -10000

In this example, no matter how the operations are taken, the left argument will be smaller than the right. So, we should just get the regular prefix-sum result: [1, 11, 111, 1111].

## Actual Behavior

It seems to take the arguments in the inverted order.
The above example yields [-10000, -10000, -10000, -10000]

## Steps to Reproduce the Problem

The code shows the problem:
```
#include <hpx/hpx_main.hpp>
#include <hpx/include/iostreams.hpp>
#include <hpx/parallel/algorithms/transform_inclusive_scan.hpp>

int main(int argc, char *argv[]) {
    std::vector<int> test{1, 10, 100, 1000};
    std::vector<int> output(test.size());
    auto pol    = hpx::parallel::execution::par(hpx::parallel::execution::task);
    auto result = hpx::parallel::transform_inclusive_scan(
                      pol, test.cbegin(), test.cend(), output.begin(),
                      [](int acc, int xs) -> int {
                          if(acc < xs) {
                              return acc + xs;
                          } else {
                              return -10000;
                          }
                      },
                      [](int el) -> int { return el; }, 0)
                      .get();
    for(auto el : output) {
        hpx::cout << el << hpx::endl;
    }
    return 0;
}
```

## Specifications

Versions:
  HPX: V1.5.0-trunk (AGAS: V3.0), Git: unknown
  Boost: V1.72.0
  Hwloc: V1.11.0

Compiling with MSVC 2019, v. 16.5.2 on Windows 10 x64.

**Comment by hkaiser** [comment]
The way I read it, the spec says that this algorithm does not assume for the operator to be commutative (see http://eel.is/c++draft/transform.inclusive.scan#4 and http://eel.is/c++draft/numeric.ops#numerics.defns-2). Please correct me if you think I read this wrongly.

**Comment by LiliumAtratum** [comment]
@hkaiser Exactly. That's why the given example should work.
One of the more common real-life examples of a non-commutative operators used in this context are:
a) string concatenation
b) segmented prefix sum

**Comment by hkaiser** [comment]
The problem with your code is not that the operator is non-commutative. The problem is that your operator is non-associative. The spec explicitly says that using such an operator may invoke non-deterministic behavior (http://eel.is/c++draft/numeric.ops#transform.inclusive.scan-8). For this reason, I'm closing this ticket as 'won't fix'. Please reopen if needed.

**Comment by LiliumAtratum** [comment]
I have chosen this simpler operator for the sake of simplicity of the example. However, if we need a truly associative operator, one can choose the segmented prefix sum operator:

```
#include <hpx/hpx_main.hpp>
#include <hpx/include/iostreams.hpp>
#include <hpx/parallel/algorithms/transform_inclusive_scan.hpp>

struct Elem {
    int value;
    bool begin;
};

int main(int argc, char *argv[]) {
    std::vector<Elem> test{Elem{1, true}, Elem{3, false}, Elem{2, true}, Elem{4, false}};
    std::vector<Elem> output(test.size());
    auto pol    = hpx::parallel::execution::par(hpx::parallel::execution::task);
    auto result = hpx::parallel::transform_inclusive_scan(
                      pol, test.cbegin(), test.cend(), output.begin(),
                      [](Elem left, Elem right) -> Elem {
                          if(right.begin) {
                              return Elem{right.value, true};
                          } else {
                              return Elem{left.value + right.value, left.begin};
                          }
                      },
                      [](Elem el) -> Elem { return el; }, Elem{0, true})
                      .get();
    for(auto el : output) {
        hpx::cout << el.value << hpx::endl;
    }
    return 0;
}
```

Intuitively the operator add up values as long as the second argument remains `false`. A value of `true` marks the beginning of a new segment and that "resets" the summation. Unless I made a mistake somewhere, I have proven myself, with pen and paper, that this operator is in fact associative.

The expected output is:
1 4 2 6
But with hpx I get:
0 1 4 2

**Comment by hkaiser** [comment]
I will need some time to think about this. For now I'll reopen the ticket so we won't forget about it.

**Comment by hkaiser** [comment]
Thanks for figuring that out. The problem will be fixed by merging #4788 as well.
